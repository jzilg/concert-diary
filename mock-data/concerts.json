[
  {
    "id": 1,
    "band": "The Cure",
    "supportBands": [
      "Talking Heads"
    ],
    "location": "C-Halle",
    "date": "2019-04-17",
    "companions": [
      "Max",
      "Peter",
      "Julia"
    ]
  },
  {
    "id": 2,
    "band": "Metallica",
    "supportBands": [
      "Tool",
      "Slayer",
      "Rammstein"
    ],
    "location": "Wuhlheide",
    "date": "2000-11-01",
    "companions": [
      "Peter",
      "Matze",
      "Leo"
    ]
  },
  {
    "id": 4,
    "band": "The Smiths",
    "supportBands": [
      "Depeche Mode"
    ],
    "location": "C-Halle",
    "date": "2019-04-01",
    "companions": []
  }
]
