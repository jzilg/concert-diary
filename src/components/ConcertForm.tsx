import React, { Fragment, useState, ReactElement } from 'react'
import { Link } from 'react-router-dom'
import Concert from '../entities/Concert.interface'
import ListInput from './ListInput'

interface Props {
    concert: Concert
    saveConcert: Function
    goToHome: Function
}

function ConcertForm(props: Props): ReactElement {
    const { concert, saveConcert, goToHome } = props

    const [act, setAct] = useState(concert.act)
    const [supportAct, setSupportAct] = useState(concert.supportAct)
    const [location, setLocation] = useState(concert.location)
    const [date, setDate] = useState(concert.date)
    const [companions, setCompanions] = useState(concert.companions)

    function handleSubmit(event): void {
        event.preventDefault()
        saveConcert({
            id: concert.id,
            act,
            supportAct,
            location,
            date,
            companions,
        })
        goToHome()
    }

    return (
        <form onSubmit={handleSubmit}>
            <label>
                <span>Band</span>
                <ListInput
                    list={act}
                    onChange={list => setAct(list)}
                    placeholder="Pink Floyd, The Cure,..."
                />
            </label>
            <label>
                <span>Support</span>
                <ListInput
                    list={supportAct}
                    onChange={list => setSupportAct(list)}
                    placeholder="The Beatles, Talking Heads,..."
                />
            </label>
            <label>
                <span>Location</span>
                <input
                    type="text"
                    value={location}
                    onChange={event => setLocation(event.target.value)}
                    placeholder="Wuhlheide"
                />
            </label>
            <label>
                <span>Date</span>
                <input
                    type="date"
                    value={date}
                    onChange={event => setDate(event.target.value)}
                />
            </label>
            <label>
                <span>Companions</span>
                <ListInput
                    list={companions}
                    onChange={list => setCompanions(list)}
                    placeholder="Leo, Max, Peter,..."
                />
            </label>
            <button type="submit">Save</button>
        </form>
    )
}

export default ConcertForm
