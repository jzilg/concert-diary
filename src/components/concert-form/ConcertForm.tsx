import React, { useEffect, FC } from 'preact/compat'
import { Link } from 'react-router-dom'
import Concert from '../../entities/Concert'
import ListInput from '../list-input'
import useFormInput from '../../hooks/useFormInput'
import useListInput from '../../hooks/useListInput'
import { FormButtonList, FormButtonListItem } from '../form-button-list'

type Props = {
    concert: Concert
    saveConcert: Function
}

const ConcertForm: FC<Props> = (props) => {
    const { concert, saveConcert } = props

    const [band, onBandInputChange, setBand] = useFormInput(concert.band)
    const [location, onLocationInputChange, setLocation] = useFormInput(concert.location)
    const [date, onDateInputChange, setDate] = useFormInput(concert.date)
    const [companions, onCompanionsInputChange, setCompanions] = useListInput(concert.companions)
    const [
        supportBands,
        onSupportBandInputChange,
        setSupportBands,
    ] = useListInput(concert.supportBands)

    useEffect(() => {
        setBand(concert.band)
        setSupportBands(concert.supportBands)
        setLocation(concert.location)
        setDate(concert.date)
        setCompanions(concert.companions)
    }, [concert]) // eslint-disable-line react-hooks/exhaustive-deps

    const handleSubmit = (event: Event): void => {
        event.preventDefault()
        saveConcert({
            id: concert.id,
            band,
            supportBands,
            location,
            date,
            companions,
        })
    }

    return (
        <form onSubmit={handleSubmit}>
            <label>
                <span>Band</span>
                <input
                    type="text"
                    value={band}
                    onChange={onBandInputChange}
                    placeholder="Pink Floyd"
                    required
                />
            </label>
            <label>
                <span>Support</span>
                <ListInput
                    list={supportBands}
                    onChange={onSupportBandInputChange}
                    placeholder="The Cure, Talking Heads,..."
                />
            </label>
            <label>
                <span>Location</span>
                <input
                    type="text"
                    value={location}
                    onChange={onLocationInputChange}
                    placeholder="Wuhlheide"
                    required
                />
            </label>
            <label>
                <span>Date</span>
                <input
                    type="date"
                    value={date}
                    onChange={onDateInputChange}
                    required
                />
            </label>
            <label>
                <span>Companions</span>
                <ListInput
                    list={companions}
                    onChange={onCompanionsInputChange}
                    placeholder="Leo, Max, Peter,..."
                />
            </label>
            <FormButtonList>
                <FormButtonListItem>
                    <button type="submit">Save</button>
                </FormButtonListItem>
                <FormButtonListItem>
                    <Link to="/concerts">Back to Concerts</Link>
                </FormButtonListItem>
            </FormButtonList>
        </form>
    )
}

export default ConcertForm
