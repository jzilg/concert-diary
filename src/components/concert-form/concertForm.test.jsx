import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import ConcertForm from './ConcertForm'
import concerts from '../../../mock-data/concerts.json'

describe('Loading', () => {
    const goToConcerts = () => {}
    const saveConcert = () => {}

    it('should render correctly', () => {
        const component = shallow((
            <ConcertForm
                concert={concerts[0]}
                goToConcerts={goToConcerts}
                saveConcert={saveConcert}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })
})
