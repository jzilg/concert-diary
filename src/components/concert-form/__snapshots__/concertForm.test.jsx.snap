// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    <span>
      Band
    </span>
    <input
      onChange={[Function]}
      placeholder="Pink Floyd"
      type="text"
      value="The Cure"
    />
  </label>
  <label>
    <span>
      Support
    </span>
    <ListInput
      list={
        Array [
          "Talking Heads",
        ]
      }
      onChange={[Function]}
      placeholder="The Cure, Talking Heads,..."
    />
  </label>
  <label>
    <span>
      Location
    </span>
    <input
      onChange={[Function]}
      placeholder="Wuhlheide"
      type="text"
      value="C-Halle"
    />
  </label>
  <label>
    <span>
      Date
    </span>
    <input
      onChange={[Function]}
      type="date"
      value="2019-04-17"
    />
  </label>
  <label>
    <span>
      Companions
    </span>
    <ListInput
      list={
        Array [
          "Max",
          "Peter",
          "Julia",
        ]
      }
      onChange={[Function]}
      placeholder="Leo, Max, Peter,..."
    />
  </label>
  <button
    type="submit"
  >
    Save
  </button>
</form>
`;
